### ISSUE TITLE
An Unexpected Segmentation Fault caused by (!()) (I guess)

### ISSUE THREAD

**Issue by haoyang9804** [ethereum/solidity#14719]
```Solidity
contract C {
    function f0() public { (()) = 2; }

    function f1() public pure { (!()) = (); }

        function f2() internal pure returns (uint, uint) { return () = f2(); }

        function f3()public{return()=();}

        function f4 ( bytes32 hash , uint8 v , bytes32 r , bytes32 s , uint blockExpired , bytes32 salt ) public returns ( address ) {
        require (
) ;
        return ecrecover ( hash , v , r , s ) ;
    }
}
```
This code triggers a segmentation fault in the latest version of solc

**Comment by haoyang9804** [comment]
This bug-triggered code can be reduced into
```Solidity
contract C {
    function f1() public pure { (!()) = (); }
}
```

**Comment by haoyang9804** [comment]
Here is another similar bug-triggered code that can also reveal a segmentation fault. I guess these two reveal the same bug
```Solidity
contract C {
  struct S { function() returns (S storage)[] x; }
  S s;
  function f() internal pure returns (uint, uint, uint, S storage, uint, uint) {
    (!f()) = f();
  }
}
```

**Comment by nikola-matic** [comment]
[lhsResolved](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/TypeChecker.cpp#L119) is `NULL` in this instance, and we're de-referencing it without checking, which is why the segfault occurs. This should not segfault regardless.

**Comment by haoyang9804** [comment]
> [lhsResolved](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/TypeChecker.cpp#L119) is `NULL` in this instance, and we're de-referencing it without checking, which is why the segfault occurs. This should not segfault regardless.

I see. Thanks for the clear explaination. Maybe I can help resolve it these days.

**Comment by nikola-matic** [comment]
Another repro taken from the soon to be closed duplicate #14720:

```Solidity
contract C
{
	function f() internal pure returns (uint, uint) {
		return (2, 3);
	}
	function g() public pure {
		uint x;
		uint y;
		(f()) = f();
	}
}
```

**Comment by nikola-matic** [comment]
Turns out this was yet another duplicate, and was fixed by https://github.com/ethereum/solidity/pull/15084. The fix is available in the 0.8.26 release.
